pkg/scene/widget_align_system.go

```diff
@@ -0,0 +1,200 @@
+type WidgetAlignSystem struct {
+	inner WidgetZone
+	outer WidgetZone
+}
+
+// WidgetZone is the structure of each layer (inner and outer) of the align system.
+type WidgetZone struct {
+	left   WidgetSection
+	center WidgetSection
+	right  WidgetSection
+}
+
+// WidgetSection is the structure of each section of the align system.
+type WidgetSection struct {
+	top    WidgetArea
+	middle WidgetArea
+	bottom WidgetArea
+}
+
+// WidgetArea has the widgets and alignment information found in each part area of a section.
+type WidgetArea struct {
+	widgetIds []*id.WidgetID
+	align     string
+}
+
+type WidgetLocation struct {
+	Zone    string
+	Section string
+	Area    string
+}
+
+// NewWidgetAlignSystem returns a new widget align system.
+func NewWidgetAlignSystem() *WidgetAlignSystem {
+	return &WidgetAlignSystem{}
+}
+
+// WidgetZone will return a specific zone in the align system.
+func (was *WidgetAlignSystem) WidgetZone(zone string) *WidgetZone {
+	if was == nil {
+		return nil
+	}
+	switch zone {
+	case "inner":
+		return &was.inner
+	case "outer":
+		return &was.outer
+	}
+	return nil
+}
+
+// WidgetSection will return a specific section in the align system.
+func (was *WidgetAlignSystem) WidgetSection(zone, section string) *WidgetSection {
+	if was == nil {
+		return nil
+	}
+
+	z := was.WidgetZone(zone)
+
+	switch section {
+	case "left":
+		return &z.left
+	case "center":
+		return &z.center
+	case "right":
+		return &z.right
+	}
+	return nil
+}
+
+// WidgetArea will return a specific area in the align system.
+func (was *WidgetAlignSystem) WidgetArea(zone, section, area string) *WidgetArea {
+	if was == nil {
+		return nil
+	}
+
+	s := was.WidgetSection(zone, section)
+
+	switch area {
+	case "top":
+		return &s.top
+	case "middle":
+		return &s.middle
+	case "bottom":
+		return &s.bottom
+	}
+	return nil
+}
+
+// WidgetIds will return a slice of widget ids from a specific area.
+func (was *WidgetAlignSystem) WidgetIds(z, s, a string) []*id.WidgetID {
+	area := was.WidgetArea(z, s, a)
+	return area.widgetIds
+}
+
+// Alignment will return the alignment of a specific area.
+func (was *WidgetAlignSystem) Alignment(z, s, a string) *string {
+	area := was.WidgetArea(z, s, a)
+	return &area.align
+}
+
+// Add a widget to the align system.
+func (was *WidgetAlignSystem) Add(wid *id.WidgetID, l *WidgetLocation) {
+	if was == nil {
+		return
+	}
+	a := was.WidgetArea(l.Zone, l.Section, l.Area)
+	nIds := a.widgetIds
+	if !a.Has(*wid) {
+		nIds = append(a.widgetIds, wid)
+	}
+	a.widgetIds = nIds
+	if a.align == "" {
+		a.align = "start"
+	}
+}
+
+// Remove a widget from the align system.
+func (was *WidgetAlignSystem) Remove(wid *id.WidgetID, l *WidgetLocation) {
+	if was == nil {
+		return
+	}
+
+	a := was.WidgetArea(l.Zone, l.Section, l.Area)
+
+	var nwid []*id.WidgetID
+	for _, w := range a.widgetIds {
+		if w.ID() != wid.ID() {
+			nwid = append(nwid, w)
+		}
+	}
```

Comment:
Instead of searching for widgets by location, search for widgets for all locations from the beginning. Otherwise, the location will behave as an ID.

In general, the information that behaves as an ID should be kept to a minimum, otherwise it will cause human errors and bugs to be introduced.

By the way, with this amount of data, you don't need to worry about the search speed at all.